plutus-core/plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -186,20 +186,67 @@ instance PrettyDefaultBy config Integer => PrettyBy config ExBudget where
         , "}"
         ]
 
-data ExBudgetState exBudgetCat = ExBudgetState
-    { _exBudgetStateTally  :: ExTally exBudgetCat -- ^ for counting what cost how much
-    , _exBudgetStateBudget :: ExBudget -- ^ for making sure we don't spend too much
-    }
+-- | @(-)@ on 'ExCPU'.
+minusExCPU :: ExCPU -> ExCPU -> ExCPU
+minusExCPU = coerce $ (-) @Integer
+
+-- | @(-)@ on 'ExMemory'.
+minusExMemory :: ExMemory -> ExMemory -> ExMemory
+minusExMemory = coerce $ (-) @Integer
+
+-- | Subtract an 'ExBudget' from an 'ExRestrictingBudget'.
+minusExBudget :: ExRestrictingBudget -> ExBudget -> ExRestrictingBudget
+ExRestrictingBudget (ExBudget cpuL memL) `minusExBudget` ExBudget cpuR memR =
+    ExRestrictingBudget $ ExBudget (cpuL `minusExCPU` cpuR) (memL `minusExMemory` memR)
+
+-- | A budgeting mode to execute an evaluator in.
+data ExBudgetMode
+    = Counting                         -- ^ For calculating the cost of execution.
+    | Tallying                         -- ^ For a detailed report on what costs how much +
+                                       --     the same overall budget that 'Counting' gives.
+    | Restricting ExRestrictingBudget  -- ^ For execution, to avoid overruns.
+
+-- | The state of budgeting. Gets initialized from an 'ExBudgetMode'.
+data ExBudgetState exBudgetCat
+    = TallyingSt (ExTally exBudgetCat) ExBudget
+    | CountingSt ExBudget
+    | RestrictingSt ExRestrictingBudget
     deriving stock (Eq, Generic, Show)
-    deriving anyclass NFData
+    deriving anyclass (NFData)
 instance ( PrettyDefaultBy config Integer, PrettyBy config exBudgetCat
          , Eq exBudgetCat, Hashable exBudgetCat, Ord exBudgetCat
          ) => PrettyBy config (ExBudgetState exBudgetCat) where
-    prettyBy config (ExBudgetState tally budget) = parens $ fold
+    prettyBy config (TallyingSt tally budget) = parens $ fold
         [ "{ tally: ", prettyBy config tally, line
         , "| budget: ", prettyBy config budget, line
         , "}"
         ]
+    prettyBy config (CountingSt budget) =
+        parens $ "required budget:" <+> prettyBy config budget <> line
+    prettyBy config (RestrictingSt budget) =
+        parens $ "final budget:" <+> prettyBy config budget <> line
+
+initExBudgetState
+    :: (Eq exBudgetCat, Hashable exBudgetCat)
+    => ExBudgetMode -> ExBudgetState exBudgetCat
+initExBudgetState Tallying             = TallyingSt mempty mempty
+initExBudgetState Counting             = CountingSt mempty
+initExBudgetState (Restricting budget) = RestrictingSt budget
+
+-- | Extract tallying info from an 'ExBudgetState' if it's 'TallyingSt'.
+-- If it's not then return an empty 'ExTally'.
+toTally
+    :: (Eq exBudgetCat, Hashable exBudgetCat)
```

Comment:
Do we need these?

Reply:
The constraints? `Data.HashMap.Monoidal` does not provide an `empty` hashmap, so we either need to use `mempty` which requires the constraints (because `(<>)` requires them) or use `empty` of `HashMap` and wrap it manually with `MonoidalHashMap` and also add `unordered-containers` to the `.cabal` file and fix `nix` (which I hate). We require those constraints in other places anyway...

... I wanted to say before I checked and found out that we in fact do not need any of the already existing `Hashable` constraints. I'll remove them all and use `HashMap.empty`, thanks for asking.

Reply:
Ah, well, we don't need them in that file, but we of course still need them downstream. Anyway, using `HashMap.empty` and doing wrapping manually seems better indeed.